WooCommerce Subscriptions compatibility: Fix the visible issue of the product data tab and meta box for some unsupported product types

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #2086

- Remove the controls' show/hide CSS classes of the product editing page from PHP side.
- Control the show or hide of the product data tab and meta box added to the product editing page by JS.
- Merge the custom-inputs script into the product-attributes script.

### Screenshots:

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/f8ef39ee-48ec-4161-b5b7-1dc8a2205794

### Detailed test instructions:

1. Install and activate WooCommerce Subscriptions and WooCommerce Product Bundles.
2. Restart the client build script (`npm start`) if it's running.
3. Go to edit a product.
4. Select each product type to see if the "Google Listings and Ads" tab and "Channel visibility" meta box are visible for supported product types and hidden for unsupported ones.
   - Supported product types:
      https://github.com/woocommerce/google-listings-and-ads/blob/f325aa1cebee24ffd95985c3dd719233e045e44a/js/src/product-attributes/index.js#L11-L18
5. Create a product with the "Simple subscription" type and go to its editing page.
6. After the page is initialized, check if the UIs mentioned in step 3 are hidden.

### Changelog entry

> Fix - WooCommerce Subscriptions compatibility: Fix the visible issue of the "Google Listings and Ads" tab and "Channel visibility" meta box for some unsupported product types.

eason9487 committed:
Remove the controls' show/hide CSS classes of the product editing page from PHP side.

eason9487 committed:
Control the show or hide of the product data tab and meta box added to the product editing page by JS.

eason9487 committed:
Merge the custom-inputs script into the product-attributes script.

eason9487 commented on src/Integration/WooCommerceSubscriptions.php:
This integration class doesn't actually participate in any runtime processing anymore now, but I'm not sure if it would be better to remove it entirely.

martynmjones replied:
What would be the downside of removing the file entirely?

eason9487 replied:
Originally I guess it might be ok to keep it as a reminder that this extension doesn't support the product types of WooCommerce Subscriptions. After second thoughts, I think it should be fine to remove it because the current way only uses an inclusive list of the supported product types. Other product types will naturally not be supported.

Removed in https://github.com/woocommerce/google-listings-and-ads/pull/2087/commits/d013eaad48069cd1110292807606541b0ca42a25

martynmjones reviewed (approved):
Hey @eason9487, thanks for your work on this issue 🙌 

Tested the branch locally with Subscriptions and Product Bundles. The GLA tab and channel visibility meta box now only appear if the product type is simple, variable, or bundle. So LGTM ✅

mikkamp commented on js/src/product-attributes/index.js:
```diff
@@ -1,4 +1,45 @@
+	'variable',
+	// Product bundle from WooCommerce Product Bundles
+	'bundle',
+] );
```
Just have one additional comment about this list. Previously we had the PHP filter `woocommerce_gla_supported_product_types` which allows an integration such as Bundles to add themselves to the list. We added this filter not only for the internal integrations we support, but also if there is a plugin which wants to add support for syncing their "specific" product type. With this list it's not possible to extend the supported product types. Can we pass this list dynamically to the JS so it's run through the `woocommerce_gla_supported_product_types` filter?

eason9487 replied:
Thanks for the suggestion!

Since the original execution order of the related processing is:
1. [AssetsHandler->add_many](https://github.com/woocommerce/google-listings-and-ads/blob/338300fc25a2b79acccd3b35a5e6f8f51175c609/src/Admin/Admin.php#L74-L75)
2. [AssetsHandler->register](https://github.com/woocommerce/google-listings-and-ads/blob/338300fc25a2b79acccd3b35a5e6f8f51175c609/src/Infrastructure/GoogleListingsAndAdsPlugin.php#L102-L105)
3. [IntegrationInitializer->register](https://github.com/woocommerce/google-listings-and-ads/blob/338300fc25a2b79acccd3b35a5e6f8f51175c609/src/Integration/IntegrationInitializer.php#L43-L46) and [WooCommerceProductBundles->init_product_types](https://github.com/woocommerce/google-listings-and-ads/blob/338300fc25a2b79acccd3b35a5e6f8f51175c609/src/Integration/WooCommerceProductBundles.php#L137-L143)
4. [AssetsHandler->enqueue_many](https://github.com/woocommerce/google-listings-and-ads/blob/338300fc25a2b79acccd3b35a5e6f8f51175c609/src/Admin/Admin.php#L77-L85)

On the first call, the inline script `glaData` is created along with the script file registration. However, the third call has not yet been made, which results in the supported product types not containing the `'bundle'`.

Therefore, I adjusted the execution order as follows in https://github.com/woocommerce/google-listings-and-ads/pull/2087/commits/d13cdf6b239f32d5261385857740834684aad0d2, so that the script registration will be called after the `IntegrationInitializer->register`.

1. [IntegrationInitializer->register](https://github.com/woocommerce/google-listings-and-ads/blob/d13cdf6b239f32d5261385857740834684aad0d2/src/Integration/IntegrationInitializer.php#L43-L46) and [WooCommerceProductBundles->init_product_types](https://github.com/woocommerce/google-listings-and-ads/blob/d13cdf6b239f32d5261385857740834684aad0d2/src/Integration/WooCommerceProductBundles.php#L137-L143)
2. [AssetsHandler->register_many](https://github.com/woocommerce/google-listings-and-ads/blob/d13cdf6b239f32d5261385857740834684aad0d2/src/Admin/Admin.php#L86): Combines the former `AssetsHandler->add_many` and `AssetsHandler->register`
3. [AssetsHandler->enqueue_many](https://github.com/woocommerce/google-listings-and-ads/blob/d13cdf6b239f32d5261385857740834684aad0d2/src/Admin/Admin.php#L87)

And removed the former calls of `AssetsHandler->add_many` and `AssetsHandler->register`.

eason9487 committed:
Remove WooCommerceSubscriptions class as it's no longer in use.

Address https://github.com/woocommerce/google-listings-and-ads/pull/2087#discussion_r1318159284

eason9487 committed:
Use the same source of the supported product types as the backend side on the front-end side.

Address https://github.com/woocommerce/google-listings-and-ads/pull/2087#discussion_r1318727583

eason9487 commented:
Hi @mikkamp, thanks for the suggestion.

As there are quite a number of changes made for https://github.com/woocommerce/google-listings-and-ads/pull/2087#discussion_r1319665596, could you take another look at this PR?

mikkamp commented on src/Assets/AssetsHandler.php:
```diff
@@ -4,14 +4,13 @@
  * @package Automattic\WooCommerce\GoogleListingsAndAds\Assets
  */
-final class AssetsHandler implements Registerable, AssetsHandlerInterface {
+final class AssetsHandler implements AssetsHandlerInterface {
```
I was a bit confused about the renaming of the add > register functions. If I recall correctly that name was previously avoided because it implemnted the Registerable interface. But I guess since you removed that here there is no issue with using the function name `register`.

eason9487 replied:
If I understood it right, only when a class implements the `Service` and `Registerable` interfaces and is tagged with its interfaces via a `AbstractServiceProvider` method like [share_with_tags](https://github.com/woocommerce/google-listings-and-ads/blob/d13cdf6b239f32d5261385857740834684aad0d2/src/Internal/DependencyManagement/AbstractServiceProvider.php#L71-L78) will be [treated as a service and registered](https://github.com/woocommerce/google-listings-and-ads/blob/d13cdf6b239f32d5261385857740834684aad0d2/src/Infrastructure/GoogleListingsAndAdsPlugin.php#L135-L140) in `GoogleListingsAndAdsPlugin`.

The previous `AssetsHandler` didn't implement `Service` and its interfaces were not tagged, so I think it should be fine to remove `Registerable` and use the register method in this way. So that it also pairs and aligns with the `enqueue` method in the perspective of `wp_register_script` and `wp_enqueue_script`.

mikkamp replied:
Thanks for the clarification. Yes that's correct that it needs to be both a `Service` and `Registerable`. The tagging isn't important as it will still fetch the classes even if it's not tagged.

I looked at all the assets further and they all seem to be registered/enqueued at the right time even after these changes. So that all looks good.

mikkamp reviewed (commented):
Thanks for all the changes. I've taken some time to look over it and so far it looks good.

But I just wanted to note that previously we would call:

`Asset::register` right after `init` hook. Now it's called directly whenever we call `AssetHandler::register` for the individual `Asset`. In the case of the collection of admin assets that should be fine because we call `register` on the hook `admin_enqueue_scripts`.

But I'd like to test that a little further also on the frontend assets and ensure there aren't any other adverse affects from changing those timings. I'll continue testing on Monday.

mikkamp commented on src/Admin/Admin.php:
```diff
@@ -152,13 +155,19 @@ protected function get_assets(): array {
+			),
 			$product_condition
-		) );
+		) )->add_inline_script( ...$gla_data_inline_script_args );
```
Based on the conditions we won't ever be on a WC Admin registered page and a (traditional) product edit page, so it's ok to pass the data to two separate scripts. I wonder though if this will always remain the case, like with the new product editing experience. Because in that case we would be setting `glaData` twice on the page.

If `applicableProductTypes` is all that's needed then why wouldn't we just pass that separately so we'd never be able to conflict:

```suggestion
		) )->add_inline_script(
			[
				'glaProductData',
				[
					'applicableProductTypes' => ProductSyncer::get_supported_product_types(),
				],
			]
		);
```

eason9487 replied:
Thanks for the suggestion.

It's uncertain if the new Product Edit page will have the case of setting `glaData` twice. Indeed, it will be better to avoid possible conflicts in advance. Separated the inline script withe different variable names in https://github.com/woocommerce/google-listings-and-ads/pull/2087/commits/0468a73eff2460467a569d0f3c29b385499b1560.

mikkamp reviewed (approved):
Thanks for all the changes. I've tested the assets and I can confirm that they are all loaded correctly with their associated data.

I just left one comment though about the data possibly conflicting. It's not a major blocker as it's not an issue right now. So I'll go ahead and approve this PR.

eason9487 committed:
Separate the inline scripts by different variable names to avoid potential conflicts in the future.

Address https://github.com/woocommerce/google-listings-and-ads/pull/2087#discussion_r1321205730
